ssh/tailssh/tailssh.go

```diff
@@ -122,19 +122,37 @@ func (srv *server) newSSHServer() (*ssh.Server, error) {
 		},
 		Version:                     "SSH-2.0-Tailscale",
 		LocalPortForwardingCallback: srv.mayForwardLocalPortTo,
-		NoClientAuthCallback: func(m gossh.ConnMetadata) (*gossh.Permissions, error) {
-			if srv.requiresPubKey(m.User(), toIPPort(m.LocalAddr()), toIPPort(m.RemoteAddr())) {
-				return nil, errors.New("public key required") // any non-nil error will do
+		NoClientAuthCallback: func(conn gossh.ConnMetadata) (*gossh.Permissions, error) {
+			ci, err := srv.connInfo(conn.User(), toIPPort(conn.LocalAddr()), toIPPort(conn.RemoteAddr()))
+			if err != nil {
+				srv.logf("failed to get conninfo: %v", err)
+				return nil, gossh.ErrDenied
+			}
+			if _, withoutPubKey := srv.canProceed(ci, nil); withoutPubKey {
+				return nil, nil
+			}
+			if srv.havePubKeyPolicy(ci) {
+				return nil, fmt.Errorf("needs more info")
 			}
-			return nil, nil
+			return nil, gossh.ErrDenied
 		},
 		PublicKeyHandler: func(ctx ssh.Context, key ssh.PublicKey) bool {
-			if srv.acceptPubKey(ctx.User(), toIPPort(ctx.LocalAddr()), toIPPort(ctx.RemoteAddr()), key) {
-				srv.logf("accepting SSH public key %s", bytes.TrimSpace(gossh.MarshalAuthorizedKey(key)))
-				return true
+			conn := ctx
+			ci, err := srv.connInfo(conn.User(), toIPPort(conn.LocalAddr()), toIPPort(conn.RemoteAddr()))
```

Comment:
this is all wasteful, doing up to 17 ssh.ConnInfo calls with the same info.

can you make this ci once early, once conn.User is first known?